src/ucp/rndv/rndv.c

```diff
@@ -969,10 +976,85 @@ ucp_rndv_send_frag_update_get_rkey(ucp_worker_h worker, ucp_request_t *freq,
     memset(rkey_index, 0, UCP_MAX_LANES * sizeof(uint8_t));
 }
 
+ucs_mpool_ops_t ucp_frag_mpool_ops = {
+    .chunk_alloc   = ucp_rndv_frag_mpool_malloc,
+    .chunk_release = ucp_rndv_frag_mpool_free,
+    .obj_init      = ucp_rndv_frag_mpool_obj_init,
+    .obj_cleanup   = ucs_empty_function
+};
+
+static ucs_mpool_t *
+ucp_rndv_get_mpool(ucp_worker_h worker, const ucp_worker_mpool_key_t *key)
+{
+    int num_frags    = 128;
+    size_t frag_size = worker->context->config.ext.rndv_frag_size[key->mem_type];
+    ucs_mpool_t *mpool;
+    khiter_t khiter;
+    int khret;
+    ucp_rndv_mpool_priv_t *mpriv;
+    ucs_status_t status;
+
+    ucs_recursive_spin_lock(&worker->mpool_hash_lock);
+
+    khiter = kh_get(ucp_worker_mpool_hash, &worker->mpool_hash, *key);
+    if (ucs_likely(khiter != kh_end(&worker->mpool_hash))) {
+        mpool = kh_val(&worker->mpool_hash, khiter);
+    } else {
+        mpool = ucs_malloc(sizeof(ucs_mpool_t), "ucp_worker_mpool");
+        status = ucs_mpool_init(mpool, sizeof(ucp_rndv_mpool_priv_t),
+                                frag_size + sizeof(ucp_mem_desc_t),
+                                sizeof(ucp_mem_desc_t),
+                                UCS_SYS_PCI_MAX_PAYLOAD, 128, UINT_MAX,
+                                &ucp_frag_mpool_ops, "ucp_rndv_frags");
+        if (status != UCS_OK) {
+            ucs_free(mpool);
+            return NULL;
+        }
+
+        mpriv              = (ucp_rndv_mpool_priv_t *)ucs_mpool_priv(mpool);
+        mpriv->worker      = worker;
+        mpriv->mem_type    = key->mem_type;
+        mpriv->num_frags   = num_frags;
+        mpriv->frag_size   = frag_size;
+        khiter             = kh_put(ucp_worker_mpool_hash, &worker->mpool_hash,
+                                    *key, &khret);
+        if (khret == UCS_KH_PUT_FAILED) {
+            ucs_mpool_cleanup(mpool, 0);
+            ucs_free(mpool);
+            return NULL;
+        }
+
+        ucs_assert_always(khret != UCS_KH_PUT_KEY_PRESENT);
+        kh_value(&worker->mpool_hash, khiter) = mpool;
+    }
+
+    ucs_recursive_spin_unlock(&worker->mpool_hash_lock);
+
+    return mpool;
+}
+
+static ucp_mem_desc_t *
+ucp_rndv_mpool_get(ucp_worker_h worker, void *buffer, size_t length,
+                   ucs_memory_type_t mem_type)
+{
+    ucp_memory_info_t mem_info;
+    ucp_worker_mpool_key_t key;
+    ucs_mpool_t *mpool;
+
+    ucp_memory_detect(worker->context, buffer, length, &mem_info);
```

Comment:
why do you need this here?

Reply:
will remove

Reply:
Eventual goal is to detect sys device id of the buffer and get a staging buffer on that device.

Reply:
another memory detection adds some overhead, at least of memtype cache locks